packages/graph/hash_graph/lib/graph/src/store/postgres/query/expression/join_clause.rs

```diff
@@ -11,6 +13,39 @@ pub struct JoinExpression<'q> {
 impl<'q> JoinExpression<'q> {
     #[must_use]
     pub const fn from_tables(table: Table, on: Table) -> Self {
+        // Crossing the boundaries of ontology <-> Knowledge requires special casing
+        match (table.name, on.name) {
+            (TableName::Entities, TableName::EntityTypes) => {
+                return Self {
+                    table,
+                    on: Condition::Equal(
+                        Some(Expression::Column(Column {
+                            table,
+                            access: ColumnAccess::Table {
+                                column: "entity_type_version_id",
+                            },
+                        })),
+                        Some(Expression::Column(on.target_join_column())),
+                    ),
+                };
+            }
+            (TableName::EntityTypes, TableName::Entities) => {
+                return Self {
+                    table,
+                    on: Condition::Equal(
+                        Some(Expression::Column(table.source_join_column())),
+                        Some(Expression::Column(Column {
+                            table: on,
+                            access: ColumnAccess::Table {
+                                column: "entity_type_version_id",
+                            },
+                        })),
+                    ),
+                };
+            }
+            _ => {}
+        }
+
         Self {
             table,
             on: Condition::Equal(
```

Comment:
nit for line 61, could we be explicit about these being `INNER JOIN`s?

Reply:
I'd like to do this in a separate PR, I added an [Asana task](https://app.asana.com/0/0/1203225514907868/f) (_internal_).